### ISSUE TITLE
Custom middlewares/interceptors

### ISSUE THREAD

**Issue by lejabque** [grpc-ecosystem/grpc-gateway#3050]
Hello, I'm trying to implement access logs in our API based on the grpc-gateway.
I want to log the request body in human-readable form (json) even if a protobuf was received.
To do this, I need some kind of middleware or interceptor in the gateway. Do I understand correctly that there is still no way to do this other than using the grpc server, as described in https://github.com/grpc-ecosystem/grpc-gateway/issues/2365? don't want to have such an overhead :(

I think this sounds like a useful thing, are there any best practices for implementing this?

**Comment by johanbrandhorst** [comment]
Hi, thanks for your issue. If you simply want to log the request body, you can implement this as a HTTP middleware wrapping the gateway mux instead of in the gRPC stack. Something like https://drstearns.github.io/tutorials/gomiddleware/ shows the general pattern. Logging the request body specifically will take a little extra work since you'll need to buffer the whole request in memory by reading the `req.Body` before passing it to the gateway handler. Hope that helps!

**Comment by lejabque** [comment]
Yes, we only provide http api without grpc and accept both jsons and serialized protos. Therefore, we do not need a grpc server except for this logging hack. But I realized that we have some exotic case and and it's wrong to change the interface for that. Anyway thanks, will use grpc server
